src/eval.c

```diff
@@ -5394,6 +5465,779 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    eval1(&string, rettv, evaluate);
+    free(origin_string);
+
+    // Consume
+    *arg += STRLEN(template);
+    free(template);
+
+    return OK;
+}
+
+/*
+ * Get a string between a opening "$'" and a closing "'" (or "$\"" and "\""),
+ * and check a taken string is not missing a closing "'".
+ *
+ * This result MUST BE `free`d later! (if it is not NULL.)
+ */
+    static char_u*
+read_template_string(int (*is_closing_quote)(char_u*), char_u *source)
+{
+    const size_t 	OPENING_QUOTE_LENGTH = 2;  // "$'"
+    char_u*		p;
+    char_u*		template;
+    size_t		template_length;
+
+    for (p = source + OPENING_QUOTE_LENGTH; !(is_closing_quote(p) || *p == NUL); MB_PTR_ADV(p));
+    if (*p == NUL)
+    {
+	semsg(_("E115: Missing quote: %s"), source);
+	return NULL;
+    }
+
+    template_length = p - source + 1;  // 1 to count by 1-based
+    template = alloc_terminated(template_length);
+    memcpy(template, source, template_length);
+    return template;
+}
+
+    static size_t
+sum_embedded_expr_strlen(embedded_t** embedded, size_t embedded_size)
+{
+    size_t	length = 0;
+    size_t	i;
+
+    for (i = 0; i < embedded_size; ++i)
+    {
+	length += STRLEN(embedded[i]->expr);
+    }
+
+    return length;
+}
+
+    static size_t
+sum_embedded_stringified_strlen(embedded_t** embedded, size_t embedded_size)
+{
+    size_t	length = 0;
+    size_t	i;
+
+    for (i = 0; i < embedded_size; ++i)
+    {
+	length += STRLEN(embedded[i]->stringified);
+    }
+
+    return length;
+}
+
+    static size_t
+sum_embedding_strlen(embedded_t** embedded, size_t embedded_size)
+{
+    const size_t	LITERAL_EMBEDDING_LENGTH = 3;  // a number of strlen("${}")
+    const size_t	VAR_EMBEDDING_LENGTH = 1;  // a number of strlen("$")
+    size_t		length = 0;
+    size_t		i;
+
+    for (i = 0; i < embedded_size; ++i)
+    {
+	length += embedded[i]->is_embedded_literal
+		? LITERAL_EMBEDDING_LENGTH
+		: VAR_EMBEDDING_LENGTH;
+    }
+
+    return length;
+}
+
+/*
+ * Count up independent "$"s on `template`.
+ * (Count up embedding.)
+ */
+     static size_t
+count_embedding(char_u *template)
+{
+    size_t	count = 0;
+    char_u*	p = template + 1;  // skip a head of '$'
+
+    // Allow to refer -1 at later
+    if (p[0] == '$' && p[1] != '$')
+    {
+	++count;
+    }
+    MB_PTR_ADV(p);
+
+    for (; *p != NUL; MB_PTR_ADV(p))
+    {
+	// Is `p` an independent '$'? e.g. "x$x". not "$$x" nor "x$$".
+	if (p[-1] != '$' && p[0] == '$' && p[1] != '$')
+	{
+	    ++count;
+	    ++p;      // a part of "$"
+	    continue; // a part of "{" (incresing)
+	}
+    }
+
+    return count;
+}
+
+/*
+ * Allocate `size + 1` memory,
+ * also set the last to NUL.
+ */
+    static char_u*
+alloc_terminated(size_t size)
+{
+    char_u*	p = alloc(size + 1);
+    p[size] = NUL;
+    return p;
+}
+
+/*
+ * Allocate a `embedded_t*`.
+ */
+    static embedded_t*
+new_embedded(int is_embedded_literal, char_u *expr, size_t expr_size)
+{
+    embedded_t *x = alloc(sizeof(embedded_t));
```

Comment:
`ALLOC_ONE` can be used.
s/alloc(sizeof(embedded_t))/ALLOC_ONE(embedded_t)/

Reply:
:+1: